### ISSUE TITLE
Allow more fine-grained suppression of `ConstructorLeaksThis` 

### ISSUE THREAD

**Issue by Stephan202** [google/error-prone#735]
Consider the following code:
```java
final class Container {
    @SuppressWarnings("ConstructorLeaksThis")
    static final class Test1 {
        Test1() {
            String s = String.valueOf(this);
        }
    }

    static final class Test2 {
        @SuppressWarnings("ConstructorLeaksThis")
        Test2() {
            String s = String.valueOf(this);
        }
    }

    static final class Test3 {
        Test3() {
            @SuppressWarnings("ConstructorLeaksThis")
            String s = String.valueOf(this);
        }
    }
}
```

Running with `-Xep:ConstructorLeaksThis:ERROR` yields:
```
Container.java:12: error: [ConstructorLeaksThis] Constructors should not pass the 'this' reference out in method invocations, since the object may not be fully constructed.
            String s = String.valueOf(this);
                                      ^
    (see http://errorprone.info/bugpattern/ConstructorLeaksThis)
Container.java:19: error: [ConstructorLeaksThis] Constructors should not pass the 'this' reference out in method invocations, since the object may not be fully constructed.
            String s = String.valueOf(this);
                                      ^
    (see http://errorprone.info/bugpattern/ConstructorLeaksThis)
2 errors
```

As you can see, no error is emitted for `Test1`. But for the other two classes `@SuppressWarnings("ConstructorLeaksThis")` has no effect.

For classes with many fields or much code in initializer blocks/constructors, it'd be nice if `ConstructorLeaksThis` could be suppressed at a scope smaller than the whole class. (Refactoring the affected classes would be even better, of course, but alas.)
